zebra: reduce impact of route-update overload

mjstapp opened the pull request:
If zebra is facing an impulse of route updates, try to reduce the memory footprint and reduce redundant work. Free redundant route_nodes promptly. Also improve the dataplane counters we show with the 'show zebra dplane provider' cli.

mjstapp commented:
Pushed to clean up SA warning

donaldsharp commented:
Memory footprint is changing significantly.  I have 1 full bgp feed, then also install 1000000 routes from sharpd and then remove them.  Before this change we have this:

eva# show mem
Memory statistics for zebra:
System allocator statistics:
  Total heap allocated:  897 MiB
  Holding block headers: 16 MiB
  Used small blocks:     0 bytes
  Used ordinary blocks:  334 MiB
  Free small blocks:     41 KiB
  Free ordinary blocks:  563 MiB
  Ordinary blocks:       1175572
  Small blocks:          494
  Holding blocks:        1
(see system documentation for 'mallinfo' for meaning)


With this change we have this:

eva# show mem
Memory statistics for zebra:
System allocator statistics:
  Total heap allocated:  976 MiB
  Holding block headers: 16 MiB
  Used small blocks:     0 bytes
  Used ordinary blocks:  334 MiB
  Free small blocks:     662 KiB
  Free ordinary blocks:  642 MiB
  Ordinary blocks:       880496
  Small blocks:          7106
  Holding blocks:        1


We need to spend more time understanding this.  I would also like to see timing installs of data pre and post this change @mjstapp

mjstapp commented:
Interesting - I don't see a difference between master and this branch, with 1M sharpd routes; here's some info.
Memory on master:
```
mjs-ubu20# sho mem
Memory statistics for zebra:
System allocator statistics:
  Total heap allocated:  555 MiB
  Holding block headers: 16 MiB
  Used small blocks:     0 bytes
  Used ordinary blocks:  430 MiB
  Free small blocks:     2272 bytes
  Free ordinary blocks:  124 MiB
  Ordinary blocks:       1426754
  Small blocks:          65
  Holding blocks:        1
```

PR:

```

mjs-ubu20# sho mem
Memory statistics for zebra:
System allocator statistics:
  Total heap allocated:  557 MiB
  Holding block headers: 16 MiB
  Used small blocks:     0 bytes
  Used ordinary blocks:  430 MiB
  Free small blocks:     2272 bytes
  Free ordinary blocks:  127 MiB
  Ordinary blocks:       1494923
  Small blocks:          65
  Holding blocks:        1
```

and the detailed memory breakdowns were just about identical too.

sharp timings were also very close:

master:
```
2020/08/04 14:10:47 SHARP: Inserting 1000000 routes
2020/08/04 14:10:59 SHARP: Installed All Items 12.601351

2020/08/04 14:14:06 SHARP: Removing 1000000 routes
2020/08/04 14:14:36 SHARP: Removed all Items 30.313193
```

PR:

```
2020/08/04 14:22:49 SHARP: Inserting 1000000 routes
2020/08/04 14:23:00 SHARP: Installed All Items 11.550441

2020/08/04 14:24:31 SHARP: Removing 1000000 routes
2020/08/04 14:24:58 SHARP: Removed all Items 27.754516
```

mjstapp commented:
rebased to fix conflicts

mjstapp commented:
CI:rerun

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/58afcbc4bb749948e9e7d18186e56f75/raw/16fb2b5d78c6834a3070b47900fcfe8f3a38da54/cr_6853_1599139017.diff | git apply
```

</p>

<p>

```diff
diff --git a/zebra/zebra_dplane.c b/zebra/zebra_dplane.c
index 53b86092b..b6a275a53 100644
--- a/zebra/zebra_dplane.c
+++ b/zebra/zebra_dplane.c
@@ -3444,9 +3444,12 @@ int dplane_show_provs_helper(struct vty *vty, bool detailed)
 		out_max = atomic_load_explicit(&prov->dp_out_max,
 					       memory_order_relaxed);
 
-		vty_out(vty, "%s (%u): in: %"PRIu64", q: %"PRIu64", q_max: %"PRIu64", out: %"PRIu64", q: %"PRIu64", q_max: %"PRIu64"\n",
-			prov->dp_name, prov->dp_id, in, in_q, in_max,
-			out, out_q, out_max);
+		vty_out(vty,
+			"%s (%u): in: %" PRIu64 ", q: %" PRIu64
+			", q_max: %" PRIu64 ", out: %" PRIu64 ", q: %" PRIu64
+			", q_max: %" PRIu64 "\n",
+			prov->dp_name, prov->dp_id, in, in_q, in_max, out,
+			out_q, out_max);
 
 		DPLANE_LOCK();
 		prov = TAILQ_NEXT(prov, dp_prov_link);
@@ -3655,10 +3658,8 @@ void dplane_provider_enqueue_out_ctx(struct zebra_dplane_provider *prov,
 	/* Maintain out-queue counters */
 	atomic_fetch_add_explicit(&(prov->dp_out_queued), 1,
 				  memory_order_relaxed);
-	curr = atomic_load_explicit(&prov->dp_out_queued,
-				    memory_order_relaxed);
-	high = atomic_load_explicit(&prov->dp_out_max,
-				    memory_order_relaxed);
+	curr = atomic_load_explicit(&prov->dp_out_queued, memory_order_relaxed);
+	high = atomic_load_explicit(&prov->dp_out_max, memory_order_relaxed);
 	if (curr > high)
 		atomic_store_explicit(&prov->dp_out_max, curr,
 				      memory_order_relaxed);
@@ -4262,8 +4263,8 @@ static int dplane_thread_loop(struct thread *event)
 	/* Locate initial registered provider */
 	prov = TAILQ_FIRST(&zdplane_info.dg_providers_q);
 
-	prov_q_counter = atomic_load_explicit(&prov->dp_in_queued,
-					      memory_order_relaxed);
+	prov_q_counter =
+		atomic_load_explicit(&prov->dp_in_queued, memory_order_relaxed);
 
 	/* Don't let the plugin queue build up too far */
 	if (prov_q_counter > 2 * limit)
diff --git a/zebra/zebra_rib.c b/zebra/zebra_rib.c
index 973535a46..26f266cac 100644
--- a/zebra/zebra_rib.c
+++ b/zebra/zebra_rib.c
@@ -2989,10 +2989,10 @@ int rib_add_multipath_nhe(afi_t afi, safi_t safi, struct prefix *p,
 
 	/* Link new re to node.*/
 	if (IS_ZEBRA_DEBUG_RIB) {
-		rnode_debug(rn, re->vrf_id,
-			    "Inserting route rn %p, re %p (%s) existing %p, same_count %d",
-			    rn, re, zebra_route_string(re->type), same,
-			    same_count);
+		rnode_debug(
+			rn, re->vrf_id,
+			"Inserting route rn %p, re %p (%s) existing %p, same_count %d",
+			rn, re, zebra_route_string(re->type), same, same_count);
 
 		if (IS_ZEBRA_DEBUG_RIB_DETAILED)
 			route_entry_dump(p, src_p, re);
@@ -3017,7 +3017,8 @@ int rib_add_multipath_nhe(afi_t afi, safi_t safi, struct prefix *p,
 				continue;
 
 			if (IS_ZEBRA_DEBUG_RIB)
-				rnode_debug(rn, re->vrf_id, "rn %p, removing unneeded re %p",
+				rnode_debug(rn, re->vrf_id,
+					    "rn %p, removing unneeded re %p",
 					    rn, re);
 
 			rib_unlink(rn, re);

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

mjstapp commented:
Rebased to newer master

mjstapp commented:
CI:rerun

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/35d71548324de8df0867c5b3d32fa537/raw/5217176c210b16e63d654fbb58b3fa410ddd6cdb/cr_6853_1603803712.diff | git apply
```

</p>

<p>

```diff
diff --git a/zebra/zebra_dplane.c b/zebra/zebra_dplane.c
index e57891d5d..83536f0f2 100644
--- a/zebra/zebra_dplane.c
+++ b/zebra/zebra_dplane.c
@@ -3483,9 +3483,12 @@ int dplane_show_provs_helper(struct vty *vty, bool detailed)
 		out_max = atomic_load_explicit(&prov->dp_out_max,
 					       memory_order_relaxed);
 
-		vty_out(vty, "%s (%u): in: %"PRIu64", q: %"PRIu64", q_max: %"PRIu64", out: %"PRIu64", q: %"PRIu64", q_max: %"PRIu64"\n",
-			prov->dp_name, prov->dp_id, in, in_q, in_max,
-			out, out_q, out_max);
+		vty_out(vty,
+			"%s (%u): in: %" PRIu64 ", q: %" PRIu64
+			", q_max: %" PRIu64 ", out: %" PRIu64 ", q: %" PRIu64
+			", q_max: %" PRIu64 "\n",
+			prov->dp_name, prov->dp_id, in, in_q, in_max, out,
+			out_q, out_max);
 
 		DPLANE_LOCK();
 		prov = TAILQ_NEXT(prov, dp_prov_link);
@@ -3694,10 +3697,8 @@ void dplane_provider_enqueue_out_ctx(struct zebra_dplane_provider *prov,
 	/* Maintain out-queue counters */
 	atomic_fetch_add_explicit(&(prov->dp_out_queued), 1,
 				  memory_order_relaxed);
-	curr = atomic_load_explicit(&prov->dp_out_queued,
-				    memory_order_relaxed);
-	high = atomic_load_explicit(&prov->dp_out_max,
-				    memory_order_relaxed);
+	curr = atomic_load_explicit(&prov->dp_out_queued, memory_order_relaxed);
+	high = atomic_load_explicit(&prov->dp_out_max, memory_order_relaxed);
 	if (curr > high)
 		atomic_store_explicit(&prov->dp_out_max, curr,
 				      memory_order_relaxed);
@@ -4298,8 +4299,8 @@ static int dplane_thread_loop(struct thread *event)
 	/* Locate initial registered provider */
 	prov = TAILQ_FIRST(&zdplane_info.dg_providers_q);
 
-	prov_q_counter = atomic_load_explicit(&prov->dp_in_queued,
-					      memory_order_relaxed);
+	prov_q_counter =
+		atomic_load_explicit(&prov->dp_in_queued, memory_order_relaxed);
 
 	/* Don't let the plugin queue build up too far */
 	if (prov_q_counter > 2 * limit)
diff --git a/zebra/zebra_rib.c b/zebra/zebra_rib.c
index 710e0b35f..d73b10648 100644
--- a/zebra/zebra_rib.c
+++ b/zebra/zebra_rib.c
@@ -3004,10 +3004,10 @@ int rib_add_multipath_nhe(afi_t afi, safi_t safi, struct prefix *p,
 
 	/* Link new re to node.*/
 	if (IS_ZEBRA_DEBUG_RIB) {
-		rnode_debug(rn, re->vrf_id,
-			    "Inserting route rn %p, re %p (%s) existing %p, same_count %d",
-			    rn, re, zebra_route_string(re->type), same,
-			    same_count);
+		rnode_debug(
+			rn, re->vrf_id,
+			"Inserting route rn %p, re %p (%s) existing %p, same_count %d",
+			rn, re, zebra_route_string(re->type), same, same_count);
 
 		if (IS_ZEBRA_DEBUG_RIB_DETAILED)
 			route_entry_dump(p, src_p, re);
@@ -3032,7 +3032,8 @@ int rib_add_multipath_nhe(afi_t afi, safi_t safi, struct prefix *p,
 				continue;
 
 			if (IS_ZEBRA_DEBUG_RIB)
-				rnode_debug(rn, re->vrf_id, "rn %p, removing unneeded re %p",
+				rnode_debug(rn, re->vrf_id,
+					    "rn %p, removing unneeded re %p",
 					    rn, re);
 
 			rib_unlink(rn, re);

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/88492a8e2ae8fcac04c92a0bc1eb99f9/raw/fba1cc3d9062a13c5a3bb0acb4aa907bc6408c48/cr_6853_1607371917.diff | git apply
```

</p>

<p>

```diff
diff --git a/zebra/zebra_dplane.c b/zebra/zebra_dplane.c
index 41ff73d77..41221ec28 100644
--- a/zebra/zebra_dplane.c
+++ b/zebra/zebra_dplane.c
@@ -3708,9 +3708,12 @@ int dplane_show_provs_helper(struct vty *vty, bool detailed)
 		out_max = atomic_load_explicit(&prov->dp_out_max,
 					       memory_order_relaxed);
 
-		vty_out(vty, "%s (%u): in: %"PRIu64", q: %"PRIu64", q_max: %"PRIu64", out: %"PRIu64", q: %"PRIu64", q_max: %"PRIu64"\n",
-			prov->dp_name, prov->dp_id, in, in_q, in_max,
-			out, out_q, out_max);
+		vty_out(vty,
+			"%s (%u): in: %" PRIu64 ", q: %" PRIu64
+			", q_max: %" PRIu64 ", out: %" PRIu64 ", q: %" PRIu64
+			", q_max: %" PRIu64 "\n",
+			prov->dp_name, prov->dp_id, in, in_q, in_max, out,
+			out_q, out_max);
 
 		DPLANE_LOCK();
 		prov = TAILQ_NEXT(prov, dp_prov_link);
@@ -3919,10 +3922,8 @@ void dplane_provider_enqueue_out_ctx(struct zebra_dplane_provider *prov,
 	/* Maintain out-queue counters */
 	atomic_fetch_add_explicit(&(prov->dp_out_queued), 1,
 				  memory_order_relaxed);
-	curr = atomic_load_explicit(&prov->dp_out_queued,
-				    memory_order_relaxed);
-	high = atomic_load_explicit(&prov->dp_out_max,
-				    memory_order_relaxed);
+	curr = atomic_load_explicit(&prov->dp_out_queued, memory_order_relaxed);
+	high = atomic_load_explicit(&prov->dp_out_max, memory_order_relaxed);
 	if (curr > high)
 		atomic_store_explicit(&prov->dp_out_max, curr,
 				      memory_order_relaxed);
diff --git a/zebra/zebra_rib.c b/zebra/zebra_rib.c
index eb6587f82..d6d6f4257 100644
--- a/zebra/zebra_rib.c
+++ b/zebra/zebra_rib.c
@@ -3028,10 +3028,10 @@ int rib_add_multipath_nhe(afi_t afi, safi_t safi, struct prefix *p,
 
 	/* Link new re to node.*/
 	if (IS_ZEBRA_DEBUG_RIB) {
-		rnode_debug(rn, re->vrf_id,
-			    "Inserting route rn %p, re %p (%s) existing %p, same_count %d",
-			    rn, re, zebra_route_string(re->type), same,
-			    same_count);
+		rnode_debug(
+			rn, re->vrf_id,
+			"Inserting route rn %p, re %p (%s) existing %p, same_count %d",
+			rn, re, zebra_route_string(re->type), same, same_count);
 
 		if (IS_ZEBRA_DEBUG_RIB_DETAILED)
 			route_entry_dump(p, src_p, re);
@@ -3056,7 +3056,8 @@ int rib_add_multipath_nhe(afi_t afi, safi_t safi, struct prefix *p,
 				continue;
 
 			if (IS_ZEBRA_DEBUG_RIB)
-				rnode_debug(rn, re->vrf_id, "rn %p, removing unneeded re %p",
+				rnode_debug(rn, re->vrf_id,
+					    "rn %p, removing unneeded re %p",
 					    rn, re);
 
 			rib_unlink(rn, re);

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

mjstapp commented:
Rebase, and include only the rib-trimming changes for now.
